### DISCUSSION TITLE
Discussion webhook events

### DISCUSSION THREAD

**Discussion by Griatch** [community/community#2943] [Discussions] [answered]
This ties into comments in #31 and answers made there, but this deserves its own request-thread.

It would be very useful to have webhooks for Discussion actions. Initially one would get far just with a `New Post/answer` event.

We, and probably many others, use external chat services (Discord/IRC in our case) as the primary support and community interaction. Our current mailing list serves as the more persistent/long-form discussion venue. We have bots echoing github PRs/commits and ML post info into the chat channels. Without webhook support, I cannot really switch over to using Discussions as it stands because without notifications, chat-users would interact only rarely with Discussions and could not be inspired to jump over to answer questions etc. It would make for two separate communities.

**Comment by evi-liu** [comment]
Thank you so much for the detailed feedback on webhooks! We definitely see the value of enabling webhooks, and are looking to bring it to y'all as soon as possible.

**Reply by ricardoboss** [reply]
Will you post updates here so we can subscibe to them?

Edit: love discussions btw ❤️

**Reply by evi-liu** [reply]
yes i will! for now, please know we're working towards a public GraphQL API and looking to follow closely with webhooks.

**Reply by XenoWarrior** [reply]
Definitely looking forward to this. 👏👏
Thanks @ricardoboss for tagging me in these threads!

**Reply by BetaWile** [reply]
Yes please come as soon as possible. 👏❤️

**Comment by ssbarnea** [comment]
Lack of events on any discussions is a real deal breaker as this prevents from exposing them to our chat channels (like irc, slack,..).

Please do add some events at least for new discussion thread.

**Reply by becca** [reply]
We're working on it!

**Reply by HDVinnie** [reply]
> We're working on it!

Glad to hear! Very much needed feature.

**Reply by alice-i-cecile** [reply]
Fantastic! This would be super useful for the [Bevy](https://github.com/bevyengine/bevy) community :)

**Comment by evi-liu** [comment]
Hey folks, which Discussions webhook events would be most valuable for you (and what do you intend to use them for)? Your feedback will help shape the first version of Discussions webhooks!

**Reply by Floffah** [reply]
for me id say just new discussions, new comments, and a comment marked as answer. i would use it as a log in a discord channel

**Reply by alice-i-cecile** [reply]
New Discussion opened, followed by new answer or reply made on a Discussion. We'd use it with Discord integration for tracking open-ended design conversations, and spotting folks asking for troubleshooting / usage help.

**Reply by HDVinnie** [reply]
> for me id say just new discussions, new comments, and a comment marked as answer. i would use it as a log in a discord channel

this

**Reply by ricardoboss** [reply]
Discord integration for logging is also our use-case. So maybe webhooks tied to a category would be useful.

**Reply by Griatch** [reply]
For me, new discussions/comments alone would be enough as a first version. As mentioned in the original post, I need to echo a link to the new content into our IRC/Discord channels.

**Comment by evi-liu** [comment]
## 🎉 Discussions webhooks private beta available

Hi folks, great news! We have a private beta available for webhooks. Here are the available webhooks:

**Discussions:**
 - `created`
 - `edited`
 - `deleted`
 - `pinned`
 - `unpinned`
 - `locked`
 - `unlocked`
 - `transferred` 
 - `category_changed` 
- `answered`
- `unanswered` 

**Discussions comments:**
 - `created`
 - `edited`
 - `deleted`

If you'd like to be help test the private beta, please respond to this thread with:
1. the user(s) to flag in
2. the repository you'll be testing on

**Reply by RoinujNosde** [reply]
User: RoinujNosde
Repo: https://github.com/RoinujNosde/SimpleClans

**Reply by ricardoboss** [reply]
Users: @ricardoboss @XenoWarrior
Repo: https://github.com/cardinal-games/age-of-aincrad.com-feedback

**Reply by alice-i-cecile** [reply]
User: @cart
Repo: https://github.com/bevyengine/bevy

**Reply by HDVinnie** [reply]
User: https://github.com/HDVinnie
Repo: https://github.com/HDInnovations/UNIT3D-Community-Edition

**Reply by prochan2** [reply]
User: https://github.com/prochan2
Repo: https://github.com/postsharp/Caravela

**Reply by Bekt** [reply]
🤚 
https://github.com/supplypike/notes (private repo)

**Reply by evi-liu** [reply]
Everyone in this thread has been flagged in and should be able to use Discussions webhooks now! Please respond with:

- what do you intend to use webhooks for?
- are there other integrations you'd need (e.g. Actions, Slack, etc)?
- any feedback or bugs on the webhooks?

**Reply by alice-i-cecile** [reply]
We're hooking it into our Discord now. Initial feedback is that the level of granularity is too low: all we actually care about is "Discussion created", or maybe "Discussion comment created". Notifications on editing and deletion feel particularly intrusive and nosiy.

**Reply by ricardoboss** [reply]
@alice-i-cecile how did you get Discord working? I only get HTTP error 400 with body `{"message": "Cannot send an empty message", "code": 50006}` in the response.

**Reply by Griatch** [reply]
@evi-liu I'm the original creator of this ticket so I would also like to join :grinning: 

User: @Griatch
Repo: https://github.com/evennia/evennia

What to use for - described in initial ticket - will use it to sync to our Discord/IRC bots when there are new messages.

**Reply by HDVinnie** [reply]
Thanks for adding me to beta. I am mainly using this for my projects Discord Community. However it does not seem to be working. 

I have 3 webhooks and 3 different discord channels to display them. One is for General Repository events which has and is still working fine. One is for GitHub Sponsors events which has never worked. Last is for new Discussions events which is also not working. Both GitHub Sponsors and Discussions webhooks show a 204 response on GitHub but dont actually show up on Discord Server channel.

Im starting to think maybe discord does not support how you send the GitHub Sponsors and Discussions webhooks or something which is a real shame since many devs use discord for their communities.

**Reply by HDVinnie** [reply]
> @alice-i-cecile how did you get Discord working? I only get HTTP error 400 with body `{"message": "Cannot send an empty message", "code": 50006}` in the response.

Make sure your webhook Content type is set to application/json

**Reply by ricardoboss** [reply]
It is set to appliction/json.

**Reply by RoinujNosde** [reply]
It's not working here. The hook gets a 204 response, yet no message is sent to Discord.

**Reply by jb3** [reply]
User: https://github.com/jb3
Repository: https://github.com/python-discord/meta

Use-case: relaying to Discord through a Cloudflare worker and taking other action based on new comments on Discussions. Integration with Discord would be nice but not required, especially since that's something Discord must implement, not GitHub.

Re: above Discord comments: It's highly likely Discord hasn't implemented this payload yet, especially since it is currently in beta. There are certain payloads that are not implemented already such as deployment statuses and package publishes, so to see it in Discord you'll need to raise that with them. Otherwise you'll need to write a web application to handle webhook payloads and forward on to Discord.

**Reply by HDVinnie** [reply]
Yeah and will be ages before they do....they still do not even support GitHub sponsor webhooks.

**Reply by jb3** [reply]
For the sponsor hooks and features that are in GA that aren't supported, it's worth raising a note with Discord at https://feedback.discord.com/, for this specific feature though there isn't much they can do at this stage.

**Reply by Bekt** [reply]
piped the events to slack via @PipedreamHQ and it looks good. 

<img width="642" alt="Screen Shot 2021-03-18 at 7 07 58 PM" src="https://user-images.githubusercontent.com/1221480/111713278-4d76ea00-881d-11eb-8c51-579af5982a4e.png">

for those curious
```
const { WebClient } = require('@slack/web-api')
const web = new WebClient(auths.slack.oauth_access_token)

const payload = steps.trigger.raw_event.body;
if (!(payload.discussion && payload.action === 'created')) {
    console.log('Not a discussion created event.');
    return;
}
return await web.chat.postMessage({
  attachments: [
        {
            "mrkdwn_in": ["text"],
            "fallback": `Discussion #${payload.discussion.number}: ${payload.discussion.title}`,
            "pretext": `New discussion started in ${payload.repository.full_name}`,
            "color": "#36a64f",
            "author_name": payload.discussion.user.login,
            "author_link": payload.discussion.user.html_url,
            "author_icon": payload.discussion.user.avatar_url,
            "title": `#${payload.discussion.number}: ${payload.discussion.title}`,
            "title_link": payload.discussion.html_url,
            "text": payload.discussion.body,
            "fields": [
                {
                    "title": "Category",
                    "value": `${payload.discussion.category.emoji} ${payload.discussion.category.name}`,
                    "short": true
                }
            ]
        }
    ],
  channel: params.channel,
  username: "GitHub Discussions",
  icon_url: "https://slack-files2.s3-us-west-2.amazonaws.com/avatars/2017-12-19/288981919427_f45f04edd92902a96859_192.png"
})
```

**Reply by rexledesma** [reply]
User: @rexledesma
Repo: https://github.com/dagster-io/dagster

**Reply by evi-liu** [reply]
✅ everyone in this thread has been flagged in and can try out Discussions webhooks now!

thank you all for the feedback. regarding Discord, we're currently in contact with their team to get the webhooks piped in properly on their end. thank you for all your patience and excitement!

**Reply by Griatch** [reply]
I have set up our custom IRCbot to handle these new events now - works like a charm so far. :+1: 
We only care about `created` discussions/comments (too spammy with edits), so we filter out every other action.

**Reply by touchaponk** [reply]
@evi-liu not sure if I can still be added to the beta list:
User: @touchaponk
Repo: https://github.com/AmityCo/Amity-Social-Cloud-Community/

**Reply by ssbarnea** [reply]
Users: @ssbarnea
Repo: https://github.com/ansible-community/ansible-lint and https://github.com/ansible-community/molecule

**Reply by timharris777** [reply]
@evi-liu, I would like to join the private beta to test discussions with organization webhooks.
**Users**: @timharris777
**Org**: cfacorp

**Reply by evi-liu** [reply]
@touchaponk and @ssbarnea, just flagged you and your repositories in ✅.

@timharris777, which repository do you want to test on?

**Reply by timharris777** [reply]
@evi-liu , is the private beta only for repo webhooks and not for org webhooks? I would love to test discussion webhooks at the organization level. If the beta is only at the repo level I can test with `cfacorp/tools-pod`. Thanks!

**Reply by ghost** [reply]
User: sayantan300
I want to try it on my organization **rovelstars**
Repo: https://github.com/rovelstars/rovel-discord-list

**Reply by mdcfe** [reply]
We'd like to test this on our repo at EssentialsX/Essentials (user: @mdcfe).

**Reply by Maldaris** [reply]
User: @Maldaris
Repo: https://github.com/Whitesands13/Whitesands

**Reply by jpadilla** [reply]
User: @jpadilla
Repo: https://github.com/Code4PuertoRico/Code4PR

**Reply by vladfrangu** [reply]
User(s): @vladfrangu @ZachyFoxx
Repo: https://github.com/DumbDogDiner/.github

**Reply by idoasisay** [reply]
User(s): @idoasisay @gotoweb @chanste @kimploo
Repo: https://github.com/codestates/agora-states

**Reply by derrod** [reply]
User(s): @derrod
Repo: https://github.com/obsproject/obs-studio

**Reply by TimGels** [reply]
User(s): @TimGels and @johanneskauffmann
Repo: https://github.com/TimGels/OneDrive-Cloud-Player

**Reply by volkantash** [reply]
User(s): @datadeveb & @erhanbaris
Repo1: https://github.com/erhanbaris/karamel
Repo2: https://github.com/turkce-yazilim-konati/yazilimca
Organization1: https://github.com/turkce-yazilim-konati

**Reply by ilaria-carlini** [reply]
User: @ilaria-carlini
Repo: https://github.com/ilaria-carlini/test-discussionGA

**Reply by Relequestual** [reply]
User: @relequestual
Repo: https://github.com/json-schema-org/community/

Intend to use for slack intergration. Webhooks will work for now.
Plan to use it to power the JSON Schema community discussions.

I’d like to roll this out to all repos in the org rather soon!

**Reply by joebudi** [reply]
User: @joebudi
Repo: https://github.com/Budibase/budibase/discussions
Reason: When a new Discussions is created, a notification is sent to a Discord channel

**Reply by webberwang** [reply]
User: @codelab-ai
Repo: https://github.com/codelab-ai/codelab.ai

**Reply by evi-liu** [reply]
Discussions webhooks are now publicly available! (see https://github.com/github/feedback/discussions/2943#discussioncomment-605964)

🎉 Just got word from the Discord team that Discord now supports our Discussions webhooks! Give it a try and feel free to share screenshots of how you're integrating Discord and Discussions together.

**Reply by ricardoboss** [reply]
@evi-liu Hey, nice to know you keep working on this feature with discord! I just tested it and unfortunately, I am still getting HTTP 400 with body `{"message": "Cannot send an empty message", "code": 50006}`. Anything I need to adjust to get it working?

**Reply by evi-liu** [reply]
@ricardoboss Hm, let me pass this along to the team to debug for you.

**Reply by HDVinnie** [reply]
working fine for me....really wish they would support your GitHub sponsor webhooks too.

**Comment by Psyruss77** [comment]
Ss

**Comment by jen-k** [comment, marked as answer]
## Discussions webhooks publicly available! 🚀 

Hi folks! Discussions webhooks are now publicly available, no need to request access. You can find more information about how to set up Discussions webhooks in the [GitHub help docs](https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads#discussion)!

Thanks for all the comments and feedback that helped us get this feature shipped!

**Reply by robaone** [reply]
This is fantastic!  Thank you!

**Reply by volkantash** [reply]
Is working with old active webhook?

**Reply by alpeer** [reply]
Discord - GitHub integration is not working(discussion feature). When will it be available?

**Reply by ilaria-carlini** [reply]
I tried using the tag `discussions` like the following:

```
on:
  discussions: 
    types: [created]
```

But apparently `discussions` is still not recognized (error: `discussions` is not a valid event name). My aim is to trigger this GitHub Action as soon as a new discussion is created. Anyone is encountering the same problem?

**Reply by reinhart1010** [reply]
Has this feature landed on Octokit yet? I would like to use this feature in [Probot](https://github.com/probot/probot) and it seems to rely on Octokit as well.

**Reply by evi-liu** [reply]
@alpeer Discord/Discussions integration should be working now! Give it a try and let me know if it's still not working for you.

@ilaria-carlini Thanks for the feedback! I'll pass it along to the team. 

@reinhart1010 Not yet, but we'll track this feature request.

**Reply by ricardoboss** [reply]
@evi-liu Discord still returns HTTP 400 for discussion events: `{"message": "Cannot send an empty message", "code": 50006}`

**Reply by volkantash** [reply]
> 
> 
> @alpeer Discord/Discussions integration should be working now! Give it a try and let me know if it's still not working for you.
> 
> @ilaria-carlini Thanks for the feedback! I'll pass it along to the team.
> 
> @reinhart1010 Not yet, but we'll track this feature request.

Discord is not work.

**Reply by evi-liu** [reply]
@ricardoboss @datadeveb Thanks for bringing this to my attention! I'm following up with the Discord team now and will keep you posted.

**Reply by ricardoboss** [reply]
@evi-liu Maybe you can also forward my method of testing discussions with Discord to their team:

1. create a Discord channel and add a webhook.
2. add the webhook url in the repository under "Settings" -> "Webhooks" -> "Add webhook" -> "Payload URL"
3. set the Content type to `application/json`
4. choose "Let me select individual events."
5. check the boxes for everything discussion related (Dicussion comments, Discussions)
6. create a new discussion or comment on an existing discussion in the repository

From the webhook settings I can see requests being sent from GitHub and the error responses from Discord mentioned above.

I hope this helps.

**Reply by volkantash** [reply]
Thanks for this info. Worked.

**Reply by JRoy** [reply]
Hello, something I noticed while testing the webhook payloads for discussions was that there was no way to tell if a discussion creation/comment came from a issue->discussion conversion. It would be great if the state property were set to `converting` or some other property was added.

**Reply by ricardoboss** [reply]
@evi-liu any updates? Still the same response...

**Reply by AshkanKD** [reply]
Thank you

**Comment by Relequestual** [comment]
For anyone interested, I've successfully created a link from GitHub Discussions to Slack via Zapier.
Here is the JSON export of the Zap https://gist.github.com/Relequestual/7684fb3df14d43b7525d9e71d44da65a

**Comment by ScottGuymer** [comment]
@evi-liu is there any plan to make similar functionality available for team discussions?
